windows: Fix playing files on network shares (#3361)

d10n opened the pull request:
gstreamer and gtk expect files on Windows network shares to begin with
file://// instead of file://

The correct uri begins with file://

This adds a utility to convert between gstreamer-compatible URIs and
correct URIs.

URI conversion is done at the point of gtk URIs coming into the
application, and at the point of correct URIs beiing sent to gstreamer.
This ensures that correct URIs are preserved internal to the
application.

Fixes #3361

Possibly related: #3351
Possibly related: #1859
Possibly related: 2895829

```
>>> GLib.filename_to_uri('\\\\192.168.1.102\\data\\Music-work\\chip.flac')
'file:////192.168.1.102/data/Music-work/chip.flac'
>>> GLib.filename_to_uri('C:\\cygwin\\setup-x86_64.exe')
'file:///C:/cygwin/setup-x86_64.exe'
>>> fsn2uri('\\\\192.168.1.102\data\Music-work\chip.flac')
'file://192.168.1.102/data/Music-work/chip.flac'
>>> fsn2uri('C:\\cygwin\\setup-x86_64.exe')
'file:///C:/cygwin/setup-x86_64.exe'
```

Check-list
----------

 * [x] There is a linked issue discussing the motivations for this feature or bugfix
 * [ ] Unit tests have been added where possible
 * [ ] I've added / updated documentation for any user-facing features.
    * N/A
 * [ ] Performance seems to be comparable or better than current `master`
    * Only tested that files play and that drag-and-drop from Explorer to playlist works

d10n commented:
Azure pipeline failure seems to be unrelated

lazka commented:
Thanks for the investigation. I'd kinda prefer if we'd fix either senf or gstreamer to do the "right" thing, whatever that may be.

lazka commented:
At least glib seems to deal fine with both variants.

lazka commented:
ah,  it only affects in case the host is an IP address, in case it is a name everything works

lazka commented on quodlibet/util/path.py:
```diff
@@ -50,6 +53,38 @@ def fsn2glib(path):
+        fsn = uri2fsn(uri)
+    except ValueError:
+        return uri
+    try:
```
Couldn't we just use the following, always?

```
from gi.repository import GLib
return GLib.filename_to_uri(uri2fsn(uri))
```

d10n replied:
GLib.filename_to_uri requires absolute paths, and I don't know if quodlibet ever uses relative paths

lazka replied:
Relative paths aren't valid in URIs, so you could do:

```
from gi.repository import GLib
try:
    return GLib.filename_to_uri(uri2fsn(uri))
except GLib.Error:
    return uri
```

lazka commented:
> Azure pipeline failure seems to be unrelated

yeah, unrelated (but something I'll have to look into :/)

d10n commented:
> Thanks for the investigation. I'd kinda prefer if we'd fix either senf or gstreamer to do the "right" thing, whatever that may be.

I thought about whether senf was the right place to put some of this, but there is no gstreamer or glib-specific stuff in senf currently, and there is other similar Windows API stuff here in quodlibet, so quodlibet seemed like the appropriate repository to put the gstreamer specific handling

d10n commented:
> Can we limit the converting to only the cases where we pass URIs to gstreamer for starters? And not the other way around and the DnD cases for example.
> 
> I'll look if I can fix senf to also handle the "////" case in uri2fsn.

I updated the DnD code because DnD did not work without the update. I tried to only convert cases that go to gstreamer and cases that come from GLib or gstreamer. If there are places I modified that doesn't need to be modified, can you point them out?

lazka commented:
Could it be DnD isn't working because `app.player.can_play_uri` which gets passed to GStreamer?

lazka commented:
I've created https://github.com/quodlibet/senf/pull/14 for senf

d10n commented:
Depends on #3372

d10n commented:
The new Azure failures might be related to the PR now

lazka commented on quodlibet/util/path.py:
```diff
@@ -50,6 +52,17 @@ def fsn2glib(path):
+        return uri
+    try:
+        # gstreamer requires extra slashes for network shares
+        return GLib.filename_to_uri(uri2fsn(uri))
```
Sorry, I missed that before,, but you also need to catch ValueError, in case non-fiel URIs get passed (like "http://etc..")

lazka replied:
I've fixed that and added some tests: 1a1284578e77a6

lazka commented:
> The new Azure failures might be related to the PR now

hm, I'll see If I can reproduce locally

lazka commented:
ok, it was just flaky

lazka commented:
thanks!

lazka commented:
I've filed this upstream now: https://gitlab.gnome.org/GNOME/glib/-/issues/2137
